nanoFramework.WebServer/HttpMultipartParser/LineReader.cs

```diff
@@ -0,0 +1,118 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.IO;
+using System.Text;
+using System.Threading;
+
+namespace nanoFramework.WebServer.HttpMultipartParser
+{
+    /// <summary>Provides methods to read a stream line by line while still returning the bytes.</summary>
+    internal class LineReader
+    {
+        private readonly Stream _stream;
+        private readonly byte[] _buffer;
+        private readonly LineBuffer _lineBuffer = new();
+        int _availableBytes = -1;
+        int _position = 0;
+
+        /// <summary>Initializes a new instance of the <see cref="LineReader"/> class.</summary>
+        /// <param name="stream">The input stream to read from.</param>
+        /// <param name="bufferSize">The buffer size to use for new buffers.</param>
+        public LineReader(Stream stream, int bufferSize)
+        {
+            _stream = stream;
+            _buffer = new byte[bufferSize];
+        }
+
+        /// <summary>
+        /// Reads a line from the stack delimited by the newline for this platform.
+        /// The newline characters will not be included in the stream.
+        /// </summary>
+        /// <param name="excludeNewLine">
+        /// If true the newline characters will be stripped when the line returns.
+        /// When reading binary data, newline characters are meaningfull and should be returned
+        /// </param>
+        /// <returns>The byte[] containing the line or null if end of stream.</returns>
+        public byte[] ReadByteLine(bool excludeNewLine = true)
+        {
+            while (true)
+            {
+                if (_position >= _availableBytes)
+                {
+                    //The stream is (should be) a NetworkStream which might still be receiving data while
+                    //we're already processing. Give the stream a chance to receive more data or we might
+                    //end up with "zero bytes read" too soon...
+                    Thread.Sleep(1);
+
+                    long streamLength = _stream.Length;
+
+                    if(streamLength > _buffer.Length)
+                    {
+                        streamLength = _buffer.Length;
+                    }
+
+                    _availableBytes = _stream.Read(_buffer, 0, (int)streamLength);
+
+                    if (_availableBytes == 0) break;
+
+                    _position = 0;
+                }
+
+                for (int i = _position; i < _availableBytes; i++)
+                {
+                    if (_buffer[i] == '\n')
+                    {
+                        //newline found, time to return the line
+                        byte[] line;
+
+                        int length = i - _position + 1;
+
+                        if (excludeNewLine)
+                        {
+                            length -= i > 0 && _buffer[i - 1] == '\r' ? 2 : 1;
```

Comment:
I guess it's more this in case of empty line:

```suggestion
                            length -= i > 1 && _buffer[i - 1] == '\r' ? 2 : 1;
```

Reply:
Not really actually, this tests when we found a \n at index i if there is a \r at index i-1. This would fail if i == 0 hence the check for i > 0